### ISSUE TITLE
Colors.{color}[x] not constant

### ISSUE THREAD

**Issue by JAicewizard** [flutter/flutter#31351]
All the sub-colours of all the colours should also be constant.

I am not sure why this isn't the case, if it can't be done then it cant be done, but it seems like something that should be done if possible for obvious reasons.

**Comment by johnsonmh** [comment]
I have thought about this as well. Based on the current implementation, I don't think it's possible.

The `Colors.red[500]` syntax works because `Colors.red` is a [`MaterialColor`](https://github.com/flutter/flutter/blob/ef276ffea501a9f9ea4d627f9d069e52a31b75be/packages/flutter/lib/src/painting/colors.dart#L432) , not just any `Color`. [`MaterialColor`](https://github.com/flutter/flutter/blob/ef276ffea501a9f9ea4d627f9d069e52a31b75be/packages/flutter/lib/src/painting/colors.dart#L432) works by extending `ColorSwatch`. `ColorSwatch` is instantiated with a Map (called `_swatch`). Since `_swatch` is an instance variable, accessing colors from the `MaterialColor` can not be done in a `const` context.  

So you can never do `const Colors.red[500]`. 

The only other similar option I can see would be defining all of the individual shades in `Colors.dart` as consts on their own.

So instead of:

```dart
  static const MaterialColor red = MaterialColor(
    _redPrimaryValue,
    <int, Color>{
       50: Color(0xFFFFEBEE),
      100: Color(0xFFFFCDD2),
      200: Color(0xFFEF9A9A),
      300: Color(0xFFE57373),
      400: Color(0xFFEF5350),
      500: Color(_redPrimaryValue),
      600: Color(0xFFE53935),
      700: Color(0xFFD32F2F),
      800: Color(0xFFC62828),
      900: Color(0xFFB71C1C),
    },
  );
  static const int _redPrimaryValue = 0xFFF44336;

```
We would do:
```dart
static const Color red50 = Color(0xFFFFEBEE);
static const Color red100 = Color(0xFFFFCDD2);
static const Color red200 = Color(0xFFEF9A9A);
static const Color red300 = Color(0xFFE57373);
static const Color red400 = Color(0xFFEF5350);
static const Color red500 = Color(_redPrimaryValue);
static const Color red600 = Color(0xFFE53935);
static const Color red700 = Color(0xFFD32F2F);
static const Color red800 = Color(0xFFC62828);
static const Color red900 = Color(0xFFB71C1C);

static const MaterialColor red = MaterialColor(
  _redPrimaryValue,
  <int, Color>{
    50: red50,
    100: red100,
    200: red200,
    300: red300,
    400: red400,
    500: red500,
    600: red600,
    700: red700,
    800: red800,
    900: red900,
  },
);
static const int _redPrimaryValue = 0xFFF44336;
```

Then, instead of doing `Colors.red[500]`, you would use, `Colors.red500` to access that shade of red in a `const` context. 

Is that something that would be helpful?

**Comment by JAicewizard** [comment]
yes, this would help. but the main issue with this would be that people might confuse this new "syntax" with that is already established with black and white ([eg](https://docs.flutter.io/flutter/material/Colors/black-constant.html))
another concern might be that many people already use red[600] and would possibly have to convert all this over, but that is an issue for whoever is going to decide if this finally gets in.

**Comment by johnsonmh** [comment]
> people might confuse this new "syntax" with that is already established with black and white (eg)

Yes, this is definitely a consideration. Maybe we could name it `redShade50`, `redShade100` ... etc. Although that might confuse users because they could also get the non `const` shades with `red.shade50`, `red.shade100` etc. 

> many people already use red[600] and would possibly have to convert all this over

What I was suggesting was to keep  the`Colors.red` AND introduce the new `Colors.red50` etc. Then we would just create the `Colors.red` out of the consts. That way you can still refer to the `MaterialColor`, but have the convenience of referring to the shades directly in a `const` context.

**Comment by JAicewizard** [comment]
Maybe introduce the new redSharexx and then depreciate the red.shadexx, I don't know the reason this was introduced but since the syntax is almost the same I can't imagine any reason to use red.shade50 over redShade50. So people will get used to the new syntax and then remove it in flutter 2 or something like that.

Having 3 different options for the same thing will confuse people who are new to flutter, you might want to use red.shade50 for personal preference and there isn't much difference compared to red[50].
But with 3 options, where one is practically an alias for another but has almost the same syntax as a third, and only the third is a constant, it might just be better to more clearly separate them.

**Comment by johnsonmh** [comment]
Well, I don't think we should deprecate red.shadexx because `red` is a `MaterialColor`, and we're not going to deprecate the `MaterialColor`s. 

I agree 3 different ways is unnecessarily complicated. 

Since 3 different methods is unnecessarily complicated and I don't think we'll deprecate `MaterialColor` anytime soon, I think we'll probably have to leave this issue as is. It's annoying that Color.{color}[x] can't be used in a `const` setting but that's a consequence of the implementation that is not easily fixed.

**Comment by RollyPeres** [comment]
Clearly all the shades should be exposed as consts. This is super poorly designed. You can't do anything useful with material color. You can't get shades as consts, you can't use static extension methods on it.

**Comment by mrienstra** [comment]
If we start with the assumption that deprecating `MaterialColor` is not an option at this time, it doesn't follow that we can't deprecate usage of `\.shade\d+` -- which only appears 121 times in the flutter codebase. (`(colors\.\w+|swatch|color)\[\d\d+\]` appears 274 times) (these searches may not be capturing everything important)

Possible roadmap:
1. Add a deprecation warning to `\.shade\d+`, and remove mentions of it from docs.
2. Add a new `const`-supporting notation (see proposals above).
3. After a suitable period of time (next breaking release?), entirely deprecate `\.shade\d+`.

End result would be the same number of methods there have now (2), without altering the most common way of using `MaterialColor` (`\[\d\d+\]`).

Unless there is a more fundamental obstacle?

Deprecating `\.shade\d+`probably wouldn't be too painful, judging by the number of search results returned by https://flutter.dev/search-all with the following search terms:

("red[50]" OR "red[100]" OR "red[200]" OR "red[300]" OR "red[400]" OR "red[500]" OR "red[600]" OR "red[700]" OR "red[800]" OR "red[900]")
--> 1870 results
("red.shade50" OR "red.shade100" OR "red.shade200" OR "red.shade300" OR "red.shade400" OR "red.shade500" OR "red.shade600" OR "red.shade700" OR "red.shade800" OR "red.shade900")
--> 41 results

("blue[50]" OR "blue[100]" OR "blue[200]" OR "blue[300]" OR "blue[400]" OR "blue[500]" OR "blue[600]" OR "blue[700]" OR "blue[800]" OR "blue[900]")
--> 1790 results
("blue.shade50" OR "blue.shade100" OR "blue.shade200" OR "blue.shade300" OR "blue.shade400" OR "blue.shade500" OR "blue.shade600" OR "blue.shade700" OR "blue.shade800" OR "blue.shade900")
--> 118 results

("grey[50]" OR "grey[100]" OR "grey[200]" OR "grey[300]" OR "grey[400]" OR "grey[500]" OR "grey[600]" OR "grey[700]" OR "grey[800]" OR "grey[900]")
--> 1350 results
("grey.shade50" OR "grey.shade100" OR "grey.shade200" OR "grey.shade300" OR "grey.shade400" OR "grey.shade500" OR "grey.shade600" OR "grey.shade700" OR "grey.shade800" OR "grey.shade900")
--> 163 results

("green[50]" OR "green[100]" OR "green[200]" OR "green[300]" OR "green[400]" OR "green[500]" OR "green[600]" OR "green[700]" OR "green[800]" OR "green[900]")
--> 1100 results
("green.shade50" OR "green.shade100" OR "green.shade200" OR "green.shade300" OR "green.shade400" OR "green.shade500" OR "green.shade600" OR "green.shade700" OR "green.shade800" OR "green.shade900")
--> 27 results

("orange[50]" OR "orange[100]" OR "orange[200]" OR "orange[300]" OR "orange[400]" OR "orange[500]" OR "orange[600]" OR "orange[700]" OR "orange[800]" OR "orange[900]")
--> 578 results
("orange.shade50" OR "orange.shade100" OR "orange.shade200" OR "orange.shade300" OR "orange.shade400" OR "orange.shade500" OR "orange.shade600" OR "orange.shade700" OR "orange.shade800" OR "orange.shade900")
--> 8 results

("yellow[50]" OR "yellow[100]" OR "yellow[200]" OR "yellow[300]" OR "yellow[400]" OR "yellow[500]" OR "yellow[600]" OR "yellow[700]" OR "yellow[800]" OR "yellow[900]")
--> 457 results
("yellow.shade50" OR "yellow.shade100" OR "yellow.shade200" OR "yellow.shade300" OR "yellow.shade400" OR "yellow.shade500" OR "yellow.shade600" OR "yellow.shade700" OR "yellow.shade800" OR "yellow.shade900")
--> 10 results

("purple[50]" OR "purple[100]" OR "purple[200]" OR "purple[300]" OR "purple[400]" OR "purple[500]" OR "purple[600]" OR "purple[700]" OR "purple[800]" OR "purple[900]")
--> 125 results
("purple.shade50" OR "purple.shade100" OR "purple.shade200" OR "purple.shade300" OR "purple.shade400" OR "purple.shade500" OR "purple.shade600" OR "purple.shade700" OR "purple.shade800" OR "purple.shade900")
--> 7 results

("pink[50]" OR "pink[100]" OR "pink[200]" OR "pink[300]" OR "pink[400]" OR "pink[500]" OR "pink[600]" OR "pink[700]" OR "pink[800]" OR "pink[900]")
--> 165 results
("pink.shade50" OR "pink.shade100" OR "pink.shade200" OR "pink.shade300" OR "pink.shade400" OR "pink.shade500" OR "pink.shade600" OR "pink.shade700" OR "pink.shade800" OR "pink.shade900")
--> 5 results

("brown[50]" OR "brown[100]" OR "brown[200]" OR "brown[300]" OR "brown[400]" OR "brown[500]" OR "brown[600]" OR "brown[700]" OR "brown[800]" OR "brown[900]")
--> 100 results
("brown.shade50" OR "brown.shade100" OR "brown.shade200" OR "brown.shade300" OR "brown.shade400" OR "brown.shade500" OR "brown.shade600" OR "brown.shade700" OR "brown.shade800" OR "brown.shade900")
--> 10 results

1870+1790+1350+1100+578+457+125+165+100 = 7,535 results
41+118+163+27+8+10+7+5+10 = 389 results
(disclaimer: while this may be an exhausting list (sorry!), it's not exhaustive, I didn't search for all of the colors names)

**Comment by arsenioo** [comment]
I would upvote this issue because it is not so difficult to generate by the script and replace all Color(0xFF....) entries used in color maps by the public named constants, like **Colors.Material.redShade500** (or similar). And at the same time keep all existing mappings and getters as is for compatibility reasons.

The most important issue with non-const material colors is that they prevent more complex classes and structures to be const like here, when **colorA** and **colorB** are not consts:

```
const BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topRight,
          end: Alignment.bottomRight,
          colors: [colorA, colorB],
        ),
        shape: BoxShape.rectangle,
        borderRadius: BorderRadius.all(
          Radius.circular(4.0),
        ),
      ),
```
